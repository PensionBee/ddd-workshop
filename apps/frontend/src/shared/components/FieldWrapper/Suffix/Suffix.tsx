import { SmallSystemIcon } from "@shared/components/Svg";
import { Box } from "@shared/components/Box";
import { BodySmallB } from "@shared/components/Typography";
import { useMemo } from "react";

type SmallSystemIconProps = React.ComponentProps<typeof SmallSystemIcon>;

export type SuffixProps = {
  /* Optional suffix icon to display i.e. AlertIcon */
  suffixIcon?: SmallSystemIconProps["icon"];
  /* Optional suffix icon fill color */
  suffixIconFill?: SmallSystemIconProps["fill"];
  /* Optional suffix text to display i.e. % */
  suffixText?: string;
};

const SuffixIcon: React.FC<SmallSystemIconProps> = (props) => (
  <SmallSystemIcon {...props} w={20} h={20} />
);

export const Suffix: React.FC<SuffixProps> = ({
  suffixIcon,
  suffixIconFill,
  suffixText,
}) => {
  const conditionalContent = useMemo(() => {
    if (suffixText) return <BodySmallB>{suffixText}</BodySmallB>;
    if (suffixIcon)
      return <SuffixIcon icon={suffixIcon} fill={suffixIconFill} />;
    return null;
  }, [suffixIcon, suffixIconFill, suffixText]);

  // Add additional padding when no suffix content
  if (!conditionalContent) return null;

  return (
    <Box d="flex" align="center" px={16} py={13} ml={-8}>
      {conditionalContent}
    </Box>
  );
};
