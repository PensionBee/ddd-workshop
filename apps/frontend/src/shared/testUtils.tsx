import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
