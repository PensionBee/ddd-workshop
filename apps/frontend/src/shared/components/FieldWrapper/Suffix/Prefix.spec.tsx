import { render } from "@shared/testUtils";
import { vi } from "vitest";
import { Suffix } from "./Suffix";

const SUFFIX_TEXT = "Suffix text";
const SUFFIX_ICON = "AlertInfo";

describe("FieldWrapper/Suffix", () => {
  describe("Text", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <Suffix suffixText={SUFFIX_TEXT} />
      );

      expect(getByText(SUFFIX_TEXT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
    it("should not render if no suffix", () => {
      const { container } = render(<Suffix />);

      expect(container.firstChild).toBeNull();
    });
  });

  describe("Icon", () => {
    it("should render correctly", () => {
      const { container, asFragment } = render(
        <Suffix suffixIcon={SUFFIX_ICON} />
      );
      const svgContainer = container.querySelector(
        `[data-rc="SmallSystemIcon.${SUFFIX_ICON}"]`
      );

      expect(svgContainer).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
    it("should not render if no suffix", () => {
      const { container } = render(<Suffix />);

      expect(container.firstChild).toBeNull();
    });
  });
});
