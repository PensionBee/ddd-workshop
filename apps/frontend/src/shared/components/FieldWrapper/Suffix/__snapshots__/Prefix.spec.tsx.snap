// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper/Suffix Icon should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR dwYaUp"
  >
    <span
      class="sc-bcXHqe lnJndv"
      data-rc="SmallSystemIcon.AlertInfo"
    >
      <svg>
        alert-info.svg
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldWrapper/Suffix Text should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR dwYaUp"
  >
    <span
      class="sc-dkrFOg lcwtgS"
      data-rc="BodySmallB"
    >
      Suffix text
    </span>
  </div>
</DocumentFragment>
`;
